src/scripts/data_file_processing/healthsystem/equipment/equipment_availability_estimation.py

```diff
@@ -0,0 +1,289 @@
+    name='Pr_Available'
+).combine_first(final_equipment_availability_export)
+
+print(f''
+      f'Fraction of missing data when requiring data for all items at all facilities = '
+      f'{round(100 * final_equipment_availability_export_full.isnull().sum() / len(final_equipment_availability_export_full))}'
+      f'%')
+
+# Impute availability of missing item_codes in a facility_id... from the mean availability of the same items in other facilities of the same level
+facility_levels = final_equipment_availability_export_full.index.get_level_values('Facility_ID').map(
+    mfl.set_index('Facility_ID')['Facility_Level'].to_dict()
+)
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby(['Item_Code', facility_levels]).transform(lambda x: x.fillna(x.mean()))
+
+# Remaining misising data include those item_codes that are never seen anywhere: interpolate these from the average availability of all other items, stratified by level and category of equipment
+final_equipment_availability_export_full = final_equipment_availability_export_full.reset_index()  # Reset level 0 index to a column for merging
+equipment_crosswalk = equipment_crosswalk[equipment_crosswalk['Item_code'].notna()]
+equipment_crosswalk['Item_Code'] = equipment_crosswalk['Item_code'].astype(int)
+final_equipment_availability_export_full = final_equipment_availability_export_full.merge(
+    equipment_crosswalk[['Item_Code', 'Category']],
+    on='Item_Code',
+    how='left',) # Merge equipment category for interpolation
+final_equipment_availability_export_full = final_equipment_availability_export_full.drop_duplicates(['Facility_ID', 'Item_Code'])
+final_equipment_availability_export_full = final_equipment_availability_export_full.set_index(['Facility_ID','Item_Code', 'Category']).squeeze()
+equipment_categories = final_equipment_availability_export_full.index.get_level_values('Category')
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby([equipment_categories, facility_levels]).transform(lambda x: x.fillna(x.mean()))
+final_equipment_availability_export_full = final_equipment_availability_export_full.droplevel(level =2) # drop equipment category from the index
+
+# Force availability to be 0 for level 5, where no service delivery occurs
+level_5_lookup = mfl.loc[mfl['Facility_Level'].isin(['5'])].set_index('Facility_Level')['Facility_ID'].to_dict()
+final_equipment_availability_export_full[final_equipment_availability_export_full.index.get_level_values(level = 0) == list(level_5_lookup.values())[0]] = 0
+
+# Remaining missing data are for items_codes in facility_ids for which there is no information at all in the facility_level
+# Impute the availability for these items in facilities as ... the average of other items available at other facilities
+final_equipment_availability_export_full = final_equipment_availability_export_full.groupby("Item_Code").transform(lambda x: x.fillna(x.mean()))
```

Comment:
I think this average should be computed from across facility levels 0-3. (Otherwise it gets dragged down by the zeros at level 5). 

I'll make that last change

Reply:
That's a good point. Thanks, Tim. Just as an FYI, there isn't any data being imputed by this line of code as the series is complete at this stage. But I have deleted the imputation in case any new source of data requires this step.